repos/system_upgrade/common/actors/checkfirstpartitionoffset/libraries/check_first_partition_offset.py

```diff
@@ -0,0 +1,96 @@
+
+
+def split_on_space_segments(line):
+    fragments = (fragment.strip() for fragment in line.split(' '))
+    return [fragment for fragment in fragments if fragment]
+
+
+def is_first_partition_offset_ok(device):
+    """
+    Check whether the offset of the first partition on the device is at least
+    1M, so that grub core fits in the gap. Expects a valid and existing GRUB device.
+    """
+    try:
+        partition_table = run(['fdisk', '-l', '-u=sectors', device], split=True)['stdout']
+    except CalledProcessError as err:
+        # Unlikely - if the disk has no partition table, `fdisk` terminates with 0 (no err). Fdisk exits with an err
+        # when the device does not exists, or if it is too small to contain a partition table.
+
+        err_msg = 'Failed to run `fdisk` to obtain the partition table of the device {0}. Full error: \'{1}\''
+        api.current_logger().error(err_msg.format(device, str(err)))
+        raise
+
+    table_iter = iter(partition_table)
+
+    for line in table_iter:
+        if not line.startswith('Units'):
+            # We are still reading general device information and not the table itself
+            continue
+
+        unit = line.split('=')[2].strip()  # Contains '512 bytes'
+        unit = int(unit.split(' ')[0].strip())
+        break  # First line of the partition table header
+
+    for line in table_iter:
+        line = line.strip()
+        if not line.startswith('Device'):
+            continue
+
+        part_all_attrs = split_on_space_segments(line)
+        break
+
+    partition_offsets = []
+    for partition_line in table_iter:
+        # Fields:               Device     Boot   Start      End  Sectors Size Id Type
+        # The line looks like: `/dev/vda1  *       2048  2099199  2097152   1G 83 Linux`
+        part_info = split_on_space_segments(partition_line)
+
+        # If the partition is not bootable, the Boot column might be empty
+        part_start = int(part_info[2]) if len(part_info) == len(part_all_attrs) else int(part_info[1])
+        partition_offsets.append(part_start)
+
+    if not partition_offsets:
+        raise ValueError('The disk {0} contains no partitions'.format(device))
+
+    first_partition_offset = min(partition_offsets)
+    offset_size = first_partition_offset * unit
+
+    return offset_size >= SAFE_OFFSET_BYTES
+
+
+def check_first_partition_offset():
+    devices_to_check = grub.get_grub_devices()
+
+    problematic_devices = []
+    for device in devices_to_check:
+        try:
+            is_ok = is_first_partition_offset_ok(device)
+            if not is_ok:
+                problematic_devices.append(device)
+        except CalledProcessError:
+            # The path does not point to an actual device, or the disk does not exists - grub.get_grub_devices lied?
+            msg = 'Failed to check the offset of the first partition of the device {0}'
+            raise StopActorExecutionError(msg.format(device))
```

Comment:
Do we need to catch `ValueError` here as well?

Reply:
note: Something to think about are systems with GPT instead of MBR. I am thinking now whether it's "safe" to raise the error when we canot do the check, as I am little bit concerned of possible false positives that could block the upgrade and increase number of customer cases. So possibly log.warning that we cannot check whether system has enough free space in MBR gap for the upgrade (or maybe even nonblocking high sev. report? informing user to verify it on their own in such a case??) would be safer solution.

Reply:
done. The `ValueError` is not raised anymore and this the code has been refactored.